cli/command_repository_upgrade.go

```diff
@@ -0,0 +1,291 @@
+		upgradeTime := l.UpgradeTime()
+		log(ctx).Infof("Repository upgrade advance notice has been set, you must come back and perform the upgrade at %s.",
+			upgradeTime)
+
+		c.skip = true
+
+		return nil
+	}
+
+	log(ctx).Infof("Repository upgrade lock intent has been placed.")
+
+	return nil
+}
+
+// drainOrCommit is the upgrade CLI phase that will actually wait for all the
+// clients to be drained out of the upgrade quorum. This phase will cause all other phases to be
+// skipped until the lock is fully established.
+func (c *commandRepositoryUpgrade) drainOrCommit(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+	cf := rep.ContentReader().ContentFormat()
+	if cf.MutableParameters.EpochParameters.Enabled {
+		log(ctx).Infof("Repository indices have already been migrated to the epoch format, no need to drain other clients")
+
+		l, err := rep.GetUpgradeLockIntent(ctx)
+		if err != nil {
+			return errors.Wrap(err, "failed to get upgrade lock intent")
+		}
+
+		if l.AdvanceNoticeDuration == 0 {
+			// let the upgrade continue to commit the new format blob
+			return nil
+		}
+
+		log(ctx).Infof("Continuing to drain since advance notice has been set")
+	}
+
+	if err := c.drainAllClients(ctx, rep); err != nil {
+		return errors.Wrap(err, "failed to upgrade the repository, lock is not released")
+	}
+	// we need to reopen the repository after this point
+
+	log(ctx).Infof("Successfully drained all repository clients, the lock has been fully-established now.")
+
+	return nil
+}
+
+// TODO(small): Better reuse.
+func sleepWithContext(ctx context.Context, dur time.Duration) {
+	t := time.NewTimer(dur)
+	defer t.Stop()
+
+	select {
+	case <-ctx.Done():
+	case <-t.C:
+	}
+}
+
+func (c *commandRepositoryUpgrade) drainAllClients(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+	password, err := c.svc.getPasswordFromFlags(ctx, false, false)
+	if err != nil {
+		return errors.Wrap(err, "getting password")
+	}
+
+	configFile, err := filepath.Abs(c.svc.repositoryConfigFileName())
+	if err != nil {
+		return errors.Wrap(err, "error resolving config file path")
+	}
+
+	lc, err := repo.LoadConfigFromFile(configFile)
+	if err != nil {
+		return errors.Wrapf(err, "error loading config file %q", configFile)
+	}
+
+	cacheOpts := lc.Caching.CloneOrDefault()
+
+	for {
+		l, err := repo.ReadAndCacheRepoUpgradeLock(ctx, rep.BlobStorage(), password, cacheOpts, -1)
+		if err != nil {
+			return errors.Wrap(err, "unable to reload the repository format blob")
+		}
+
+		upgradeTime := l.UpgradeTime()
+		now := rep.Time()
+
+		log(ctx).Infof("Waiting for %s to allow all other kopia clients to drain ...", upgradeTime.Sub(rep.Time()).Round(time.Second))
+
+		locked, writersDrained := l.IsLocked(now)
+		if locked {
+			if writersDrained {
+				// we have the lock now
+				break
+			}
+		} else {
+			return errors.Wrap(err, "upgrade lock got revoked after the intent was placed, giving up")
+		}
+
+		// TODO: this can get stuck
```

Comment:
why? how?

Reply:
on Macs and systems which have resume from sleep capability, certain time jumps could cause a longer than expected wait time here